src/wasm/wasm-binary.cpp

```diff
@@ -2386,8 +2411,10 @@ void WasmBinaryBuilder::pushBlockElements(Block* curr,
   assert(start <= expressionStack.size());
   // The results of this block are the last values pushed to the expressionStack
   Expression* results = nullptr;
-  if (type != Type::none) {
+  if (type.isSingle()) {
     results = popNonVoidExpression();
+  } else if (type.isMulti()) {
+    results = getTuple(type.size());
   }
```

Comment:
this pattern of popNonVoid-or-getTuple recurs a few times, maybe it can be refactored into a function? (which would be called if the type is none or the arity > 0 etc.)